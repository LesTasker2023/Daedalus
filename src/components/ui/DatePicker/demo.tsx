'use client';

import { useState } from 'react';
import { DatePicker } from '.';

export default function DatePickerDemo() {
  const [date, setDate] = useState<Date | undefined>(new Date());
  const [secondDate, setSecondDate] = useState<Date | undefined>();

  return (
    <div className="space-y-6 p-6 max-w-2xl mx-auto">
      <div className="space-y-2">
        <h1 className="text-2xl font-bold">DatePicker Component</h1>
        <p className="text-muted-foreground">Date selection component</p>
      </div>

      <div className="space-y-4">
        <div className="space-y-2">
          <h3 className="text-lg font-semibold">Basic Usage</h3>
          <div className="border rounded-lg p-4 bg-background">
            <DatePicker date={date} onDateChange={setDate} />
            <p className="text-sm text-muted-foreground mt-2">
              Selected: {date?.toLocaleDateString() || 'None'}
            </p>
          </div>
        </div>

        <div className="space-y-2">
          <h3 className="text-lg font-semibold">Without Default Date</h3>
          <div className="border rounded-lg p-4 bg-background space-y-2">
            <DatePicker date={secondDate} onDateChange={setSecondDate} placeholder="Pick a date" />
            <p className="text-sm text-muted-foreground">
              Selected: {secondDate?.toLocaleDateString() || 'None'}
            </p>
          </div>
        </div>
      </div>
    </div>
  );
}
