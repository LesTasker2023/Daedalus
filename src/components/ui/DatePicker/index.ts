export * from './DatePicker';
export { default as DatePickerDemo } from './demo';
