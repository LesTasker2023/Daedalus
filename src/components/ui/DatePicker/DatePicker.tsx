'use client';

import { format } from 'date-fns';
import { Calendar as CalendarIcon } from 'lucide-react';
import * as React from 'react';

import Button from '@/components/ui/Button/button';
import { Calendar } from '@/components/ui/Calendar';
import { Popover, PopoverContent, PopoverTrigger } from '@/components/ui/Popover/PopoverReal';
import { cn } from '@/lib/utils';

export interface DatePickerProps {
  date?: Date;
  onDateChange?: (date: Date | undefined) => void;
  placeholder?: string;
  disabled?: boolean;
  className?: string;
  disabledDates?: (date: Date) => boolean;
  dateFormat?: string;
}

const DatePicker = React.forwardRef<HTMLButtonElement, DatePickerProps>(
  (
    {
      date,
      onDateChange,
      placeholder = 'Pick a date',
      disabled = false,
      className,
      disabledDates,
      dateFormat = 'PPP',
      ...props
    },
    ref
  ) => {
    const [open, setOpen] = React.useState(false);

    return (
      <Popover open={open} onOpenChange={setOpen}>
        <PopoverTrigger asChild>
          <Button
            ref={ref}
            variant="outline"
            disabled={disabled}
            className={cn(
              'w-full justify-start text-left font-normal',
              !date && 'text-muted-foreground',
              className
            )}
            {...props}
          >
            <CalendarIcon className="mr-2 h-4 w-4" />
            {date ? format(date, dateFormat) : <span>{placeholder}</span>}
          </Button>
        </PopoverTrigger>
        <PopoverContent className="w-auto p-0" align="start">
          <Calendar
            mode="single"
            selected={date}
            onSelect={selectedDate => {
              onDateChange?.(selectedDate);
              setOpen(false);
            }}
            disabled={disabledDates}
            initialFocus
          />
        </PopoverContent>
      </Popover>
    );
  }
);

DatePicker.displayName = 'DatePicker';

export { DatePicker };
