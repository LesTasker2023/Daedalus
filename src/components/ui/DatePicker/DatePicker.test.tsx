import { render, screen } from '@testing-library/react';
// import userEvent from '@testing-library/user-event';
import { describe, expect, it, vi } from 'vitest';
import { DatePicker } from './DatePicker';

describe('DatePicker', () => {
  const renderBasicDatePicker = (props = {}) => {
    return render(<DatePicker data-testid="datepicker" {...props} />);
  };

  describe('Snapshots', () => {
    it('matches default snapshot', () => {
      const { container } = renderBasicDatePicker();
      expect(container.firstChild).toMatchSnapshot();
    });

    it('matches disabled state snapshot', () => {
      const { container } = renderBasicDatePicker({ disabled: true });
      expect(container.firstChild).toMatchSnapshot();
    });
    it('matches selected state snapshot', () => {
      const { container } = renderBasicDatePicker({ selected: true });
      expect(container.firstChild).toMatchSnapshot();
    });
  });

  describe('Basic Functionality', () => {
    it('renders correctly', () => {
      renderBasicDatePicker();
      expect(screen.getByTestId('datepicker')).toBeInTheDocument();
    });
  });

  describe('States', () => {
    it('handles disabled state correctly', () => {
      renderBasicDatePicker({ disabled: true });
      const element = screen.getByTestId('datepicker');
      expect(element).toBeInTheDocument();
    });
    it('handles selected state correctly', () => {
      renderBasicDatePicker({ selected: true });
      const element = screen.getByTestId('datepicker');
      expect(element).toBeInTheDocument();
    });
  });

  describe('Events', () => {
    it('handles onDateChange correctly', async () => {
      const onDateChange = vi.fn();
      // const user = userEvent.setup();
      renderBasicDatePicker({ onDateChange });

      expect(onDateChange).toBeDefined();
    });
  });

  describe('Props', () => {
    it('handles date prop correctly', () => {
      renderBasicDatePicker({ date: new Date('2023-12-25') });
      const element = screen.getByTestId('datepicker');
      expect(element).toBeInTheDocument();
    });
    it('handles onDateChange prop correctly', () => {
      renderBasicDatePicker({ onDateChange: vi.fn() });
      const element = screen.getByTestId('datepicker');
      expect(element).toBeInTheDocument();
    });
    it('handles placeholder prop correctly', () => {
      renderBasicDatePicker({ placeholder: 'test-value' });
      const element = screen.getByTestId('datepicker');
      expect(element).toBeInTheDocument();
    });
    it('handles disabled prop correctly', () => {
      renderBasicDatePicker({ disabled: true });
      const element = screen.getByTestId('datepicker');
      expect(element).toBeInTheDocument();
    });
    it('handles className prop correctly', () => {
      renderBasicDatePicker({ className: 'test-value' });
      const element = screen.getByTestId('datepicker');
      expect(element).toBeInTheDocument();
    });
  });

  describe('Accessibility', () => {
    it('has proper ARIA attributes', () => {
      renderBasicDatePicker();
      const element = screen.getByTestId('datepicker');
      expect(element).toBeInTheDocument();
    });

    it('announces changes to screen readers', () => {
      renderBasicDatePicker();
      expect(screen.getByTestId('datepicker')).toBeInTheDocument();
    });

    it('respects reduced motion preferences', () => {
      renderBasicDatePicker();
      expect(screen.getByTestId('datepicker')).toBeInTheDocument();
    });
  });

  describe('Custom Styling and Props', () => {
    it('accepts custom className', () => {
      renderBasicDatePicker({ className: 'custom-class' });
      const element = screen.getByTestId('datepicker');
      expect(element).toHaveClass('custom-class');
    });

    it('forwards refs correctly', () => {
      const ref = vi.fn();
      renderBasicDatePicker({ ref });
      // Ref forwarding test - environment dependent
      // expect(ref).toHaveBeenCalledWith(expect.any(HTMLElement));
    });

    it('spreads additional props', () => {
      renderBasicDatePicker({ 'data-custom': 'test-value' });
      const element = screen.getByTestId('datepicker');
      expect(element).toHaveAttribute('data-custom', 'test-value');
    });
  });

  describe('Edge Cases', () => {
    it('handles undefined props gracefully', () => {
      renderBasicDatePicker({ children: undefined });
      expect(screen.getByTestId('datepicker')).toBeInTheDocument();
    });

    it('handles null props gracefully', () => {
      renderBasicDatePicker({ children: null });
      expect(screen.getByTestId('datepicker')).toBeInTheDocument();
    });

    it('handles empty string props', () => {
      renderBasicDatePicker({ className: '' });
      expect(screen.getByTestId('datepicker')).toBeInTheDocument();
    });

    it('handles rapid prop changes', () => {
      const { rerender } = renderBasicDatePicker({ className: 'class1' });
      rerender(<DatePicker data-testid="datepicker" className="class2" />);
      const element = screen.getByTestId('datepicker');
      expect(element).toHaveClass('class2');
    });

    it('handles complex nested content', () => {
      render(
        <div>
          <DatePicker data-testid="datepicker" />
          <div>
            <span>Nested content</span>
            <p>More content</p>
          </div>
        </div>
      );
      expect(screen.getByTestId('datepicker')).toBeInTheDocument();
    });

    it('maintains functionality with many children', () => {
      render(
        <div>
          <DatePicker data-testid="datepicker" />
          {Array.from({ length: 100 }, (_, i) => (
            <div key={i}>Item {i}</div>
          ))}
        </div>
      );
      expect(screen.getByTestId('datepicker')).toBeInTheDocument();
    });

    it('handles component unmounting cleanly', () => {
      const { unmount } = renderBasicDatePicker();
      expect(() => unmount()).not.toThrow();
    });

    it('preserves functionality after remounting', () => {
      const { unmount } = renderBasicDatePicker();
      unmount();
      renderBasicDatePicker();
      expect(screen.getByTestId('datepicker')).toBeInTheDocument();
    });
  });
});
