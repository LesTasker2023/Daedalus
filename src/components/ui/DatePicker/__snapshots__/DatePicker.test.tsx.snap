// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`DatePicker > Snapshots > matches default snapshot 1`] = `
<button
  aria-controls="radix-«r0»"
  aria-expanded="false"
  aria-haspopup="dialog"
  class="inline-flex items-center gap-2 whitespace-nowrap rounded-md text-sm transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg]:size-4 [&_svg]:shrink-0 border border-input bg-background shadow-sm hover:bg-accent hover:text-accent-foreground [color:hsl(var(--surface-interactive))] [border-color:hsl(var(--surface-interactive))] h-9 px-4 py-2 button w-full justify-start text-left font-normal text-muted-foreground"
  data-state="closed"
  data-testid="datepicker"
  type="button"
>
  <svg
    aria-hidden="true"
    class="lucide lucide-calendar mr-2 h-4 w-4"
    fill="none"
    height="24"
    stroke="currentColor"
    stroke-linecap="round"
    stroke-linejoin="round"
    stroke-width="2"
    viewBox="0 0 24 24"
    width="24"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M8 2v4"
    />
    <path
      d="M16 2v4"
    />
    <rect
      height="18"
      rx="2"
      width="18"
      x="3"
      y="4"
    />
    <path
      d="M3 10h18"
    />
  </svg>
  <span>
    Pick a date
  </span>
</button>
`;

exports[`DatePicker > Snapshots > matches disabled state snapshot 1`] = `
<button
  aria-controls="radix-«r1»"
  aria-expanded="false"
  aria-haspopup="dialog"
  class="inline-flex items-center gap-2 whitespace-nowrap rounded-md text-sm transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg]:size-4 [&_svg]:shrink-0 border border-input bg-background shadow-sm hover:bg-accent hover:text-accent-foreground [color:hsl(var(--surface-interactive))] [border-color:hsl(var(--surface-interactive))] h-9 px-4 py-2 button w-full justify-start text-left font-normal text-muted-foreground"
  data-state="closed"
  data-testid="datepicker"
  disabled=""
  type="button"
>
  <svg
    aria-hidden="true"
    class="lucide lucide-calendar mr-2 h-4 w-4"
    fill="none"
    height="24"
    stroke="currentColor"
    stroke-linecap="round"
    stroke-linejoin="round"
    stroke-width="2"
    viewBox="0 0 24 24"
    width="24"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M8 2v4"
    />
    <path
      d="M16 2v4"
    />
    <rect
      height="18"
      rx="2"
      width="18"
      x="3"
      y="4"
    />
    <path
      d="M3 10h18"
    />
  </svg>
  <span>
    Pick a date
  </span>
</button>
`;

exports[`DatePicker > Snapshots > matches selected state snapshot 1`] = `
<button
  aria-controls="radix-«r2»"
  aria-expanded="false"
  aria-haspopup="dialog"
  class="inline-flex items-center gap-2 whitespace-nowrap rounded-md text-sm transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg]:size-4 [&_svg]:shrink-0 border border-input bg-background shadow-sm hover:bg-accent hover:text-accent-foreground [color:hsl(var(--surface-interactive))] [border-color:hsl(var(--surface-interactive))] h-9 px-4 py-2 button w-full justify-start text-left font-normal text-muted-foreground"
  data-state="closed"
  data-testid="datepicker"
  type="button"
>
  <svg
    aria-hidden="true"
    class="lucide lucide-calendar mr-2 h-4 w-4"
    fill="none"
    height="24"
    stroke="currentColor"
    stroke-linecap="round"
    stroke-linejoin="round"
    stroke-width="2"
    viewBox="0 0 24 24"
    width="24"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M8 2v4"
    />
    <path
      d="M16 2v4"
    />
    <rect
      height="18"
      rx="2"
      width="18"
      x="3"
      y="4"
    />
    <path
      d="M3 10h18"
    />
  </svg>
  <span>
    Pick a date
  </span>
</button>
`;
